### PR TITLE
G2P CTC - allow setting vocab_file from config

### PR DESCRIPTION
**Pull request by ppisljar**
# What does this PR do ?

ctc G2P model always generates the vocab_file which is only suitable for english. This allows you to set custom vocab_file thru config.

**Collection**: TTS/G2p

# Changelog 
- allow passing custom vocab file to grapheme_tokenizer

### CODE REVIEW

**Comment by ppisljar** [comment]
jenkins

**Comment by ppisljar** [comment]
@titu1994 what should i do to move this forward ? thank you.

**Comment by titu1994** [comment]
G2P would be @rlangman 
Plus, we cannot write to file inside of training step or val step. File writing must be done offline somewhere else cause ddp will introduce too many cases to it.

**Comment by ppisljar** [comment]
file writing already happened here, i didnt change anything about it, except i skip file writing if dict is provided, so i guess for DDP this should work if file is provided and not generated, so i think this should be an improvement for tht case as well

**Review by ekmb — APPROVED** [review]

**Comment by ekmb** [comment]
jenkins

File: nemo/collections/tts/g2p/models/ctc.py [hunk 1]
```diff
@@ -125,10 +125,14 @@ def setup_grapheme_tokenizer(self, cfg):
...
-            vocab_file = "/tmp/char_vocab.txt"
-            with open(vocab_file, "w") as f:
-                [f.write(f'"{ch}"\n') for ch in chars]
-                f.write('"\\""\n')  # add " to the vocab
+            if cfg.tokenizer_grapheme.vocab_file:
+                vocab_file = cfg.tokenizer_grapheme.vocab_file
+            else:
+                vocab_file = "/tmp/char_vocab.txt"
+
+                with open(vocab_file, "w") as f:
```
Comment (ekmb) [inline nemo/collections/tts/g2p/models/ctc.py:133]: we actually do need to have this for a proper model restoration [tests are failing], could you move lines 133-135 outside of `else`?

**Review by ekmb — CHANGES_REQUESTED** [review]

File: nemo/collections/tts/g2p/models/ctc.py (hunk 1 above)
Comment (ppisljar) [reply nemo/collections/tts/g2p/models/ctc.py:131]: shouldn't restoration work as the tokenizer_grapheme.vocab_file is not set ? i will look into it,
but just moving the lines out of if file would make the if do nothing.

**Comment by ppisljar** [comment]
i hope you can reopen this, i was still waiting for the review, as far as i am aware this is ready to merge

**Comment by ekmb** [comment]
jenkins

**Comment by ekmb** [comment]
jenkins

**Comment by ekmb** [comment]
jenkins

**Comment by ekmb** [comment]
jenkins

**Comment by ekmb** [comment]
@ppisljar [this test](https://github.com/NVIDIA/NeMo/blob/main/Jenkinsfile#L1216) is still failing

**Pull request closed without merging**